server/src/main/java/org/opensearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -316,60 +316,85 @@ && isTargetSameHistory()
-                 * make sure to do this before sampling the max sequence number in the next step, to ensure that we send
-                 * all documents up to maxSeqNo in phase2.
-                 */
-                RunUnderPrimaryPermit.run(
-                    () -> shard.initiateTracking(request.targetAllocationId()),
-                    shardId + " initiating tracking of " + request.targetAllocationId(),
-                    shard,
-                    cancellableThreads,
-                    logger
-                );
-
-                final long endingSeqNo = shard.seqNoStats().getMaxSeqNo();
-                if (logger.isTraceEnabled()) {
-                    logger.trace("snapshot translog for recovery; current size is [{}]", countNumberOfHistoryOperations(startingSeqNo));
-                }
-                final Translog.Snapshot phase2Snapshot = shard.newChangesSnapshot(
-                    PEER_RECOVERY_NAME,
-                    startingSeqNo,
-                    Long.MAX_VALUE,
-                    false,
-                    true
-                );
-                resources.add(phase2Snapshot);
-                retentionLock.close();
+            boolean isRecoveringReplicaWithRemoteTxLogEnabledIndex = request.isPrimaryRelocation() == false
+                && shard.isRemoteTranslogEnabledOnPrimary();
+
+            if (isRecoveringReplicaWithRemoteTxLogEnabledIndex) {
+                sendFileStep.whenComplete(r -> {
+                    assert Transports.assertNotTransportThread(RecoverySourceHandler.this + "[prepareTargetForTranslog]");
+                    // For a sequence based recovery, the target can keep its local translog
+                    prepareTargetForTranslog(0, prepareEngineStep);
+                }, onFailure);
+
+                prepareEngineStep.whenComplete(prepareEngineTime -> {
+                    assert Transports.assertNotTransportThread(RecoverySourceHandler.this + "[phase2]");
+                    RunUnderPrimaryPermit.run(
+                        () -> shard.initiateTracking(request.targetAllocationId()),
+                        shardId + " initiating tracking of " + request.targetAllocationId(),
+                        shard,
+                        cancellableThreads,
+                        logger
+                    );
+                    final long endingSeqNo = shard.seqNoStats().getMaxSeqNo();
+                    retentionLock.close();
+                    sendSnapshotStep.onResponse(new SendSnapshotResult(endingSeqNo, 0, TimeValue.ZERO));
+                }, onFailure);
+            } else {
+                sendFileStep.whenComplete(r -> {
+                    assert Transports.assertNotTransportThread(RecoverySourceHandler.this + "[prepareTargetForTranslog]");
+                    // For a sequence based recovery, the target can keep its local translog
+                    prepareTargetForTranslog(countNumberOfHistoryOperations(startingSeqNo), prepareEngineStep);
```

Comment:
what does this mean for remote translog and segrep on replica? is this a no-op?

Reply:
this method on the target triggers `openEngineAndSkipTranslogRecovery` on the engine along with updating checkpoint on replication tracker. The plan is to introduce a replication tracker no-op proxy. However, we still need the engine to be opened up.